app/src/main/java/com/aletheiaware/perspective/android/WorldAdapter.java

```diff
@@ -102,31 +109,77 @@ public void onClick(View v) {
     }
 
     @Override
-    public void onBindViewHolder(@NonNull WorldViewHolder holder, int position) {
-        if (names.isEmpty()) {
-            holder.setEmptyView();
+    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
+        if (position == 0) {
+            StarCountHolder sch = (StarCountHolder) holder;
+            int earned = 0;
+            int max = 0;
+            for (String name : names) {
+                Integer i = starsMap.get(name);
+                if (i != null) {
+                    earned += i;
+                }
+                World w = worldsMap.get(name);
+                if (w != null) {
+                    max += w.getPuzzleCount() * PerspectiveAndroidUtils.MAX_STARS;
+                }
+            }
+            sch.setStars(earned, max);
         } else {
-            String name = names.get(position);
-            World world = worldsMap.get(name);
-            if (world == null) {
-                holder.set(name, pricesMap.get(name));
+            WorldViewHolder wvh = (WorldViewHolder) holder;
+            if (names.isEmpty()) {
+                wvh.setEmptyView();
             } else {
-                int s = 0;
-                Integer stars = starsMap.get(name);
-                if (stars != null) {
-                    s = stars;
+                String name = names.get(position - 1);
+                World world = worldsMap.get(name);
+                if (world == null) {
+                    wvh.set(name, pricesMap.get(name));
+                } else {
```

Comment:
Questions from being ignorant:

- There is a name, but the world doesn't exist so its price is set, seems odd?
- What is a world's "price"?
- How do we know the price is valid?

Reply:
Initially all the free worlds are added to this adapter, so there is no price set and the world object is not null. Then Play's In-App Billing Library is queried for all the paid worlds which returns the price for each world and whether the player has purchased the world. If they have purchased it they can play it, if not the price is displayed and the player can click it to purchase it.

Reply:
What happens if the pricesMap initialization was refactored and the `pricesMap.get(name)` return null?

Reply:
I'm not sure I understand "pricesMap initialization was refactored".

A world can be added to the adapter in two ways;
- addWorld(world, stars) - the world is playable, price is null and not used - free worlds and paid worlds which the player has purchased
- addWorld(name, price) - the world is for sale, price is not null and used - paid worlds which the player has not purchased

Reply:
By refactoring I meant its possible `prices.get(name)` could be null. For example initializing prices came after adding the notify:
```
    public synchronized void addWorld(String name, String price) {
        if (!names.contains(name)) {
            names.add(name);
        }
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                notifyDataSetChanged();
            }
        });
        pricesMap.put(name, price);
    }
```
The above is DUMB, but with my ignorance, I see no obvious connection between the noitfyDataSetChanged and onBindViewHolder so I WILL do something DUMB. :)

Anyway, in onBindViewHolder you check for `names.isEmpty` and `starsMap.get(name) != null` but there is no `pricesMap.get(name) != null`. The upshot is, currently there is no problem, but unexpected things could happen in the future. :) One solution could be to add an `assert` so it fails fast?

Your call.

Reply:
Ah ok I see your point, nice catch Wink.

I've added a null check to the price and in the unlikely event it is null it will be converted to "?" - which admittedly isn't very helpful but certainly better than the app crashing. If the player taps on "?" it will open Play Billing which should have the correct price.